<a href="https://colab.research.google.com/github/Andreina-A/Data620/blob/main/Data620Project3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Andreina Arias
<br>Project 3

Using any of the three classifiers described in chapter 6 of Natural Language
Processing with Python, and any features you can think of, build the best name
gender classifier you can. Begin by splitting the Names Corpus into three
subsets: 500 words for the test set, 500 words for the dev-test set, and the
remaining 6900 words for the training set. Then, starting with the example name
gender classifier, make incremental improvements. Use the dev-test set to check
your progress. Once you are satisfied with your classifier, check its final
performance on the test set. How does the performance on the test set compare
to the performance on the dev-test set? Is this what you'd expect?

Source: Natural Language Processing with Python, exercise 6.10.2.

In [15]:
import pandas as pd
import random
import nltk
nltk.download('names') # downloaded the 'names' corpus
from nltk.corpus import names

[nltk_data] Downloading package names to /root/nltk_data...
[nltk_data]   Package names is already up-to-date!


In [16]:
#Prepare a list of examples and corresponding class labels
labeled_names = ([(name, 'male') for name in names.words('male.txt')] +
[(name, 'female') for name in names.words('female.txt')])
random.seed(28) #this ensures the shuffle is always the same
random.shuffle(labeled_names)

In [17]:
print(names.words()[:10])

['Abagael', 'Abagail', 'Abbe', 'Abbey', 'Abbi', 'Abbie', 'Abby', 'Abigael', 'Abigail', 'Abigale']


In [18]:
print(labeled_names.__len__())
print(labeled_names[:10])

7944
[('Ferdy', 'male'), ('Biddy', 'female'), ('Sibyl', 'female'), ('Thurstan', 'male'), ('Antonin', 'male'), ('Levin', 'male'), ('Zahara', 'female'), ('Nari', 'female'), ('Alberta', 'female'), ('Quintilla', 'female')]


In [19]:
# split the Names Corpus into three subsets: 500 words for the test set, 500 words for the dev-test set, and the remaining 6900 words for the training set
test_names = labeled_names[:500]
devtest_names = labeled_names[500:1000]
train_names = labeled_names[1000:]

**Features**

In [20]:
# created a baseline feature extractor (last letter)
def gender_features(word):
    return {'last_letter': word[-1].lower()}

#converted to feature sets
train_set = [(gender_features(n), g) for (n, g) in train_names]
devtest_set = [(gender_features(n), g) for (n, g) in devtest_names]
test_set = [(gender_features(n), g) for (n, g) in test_names]

#obtained accuracy using naive Bayes classifier
classifier=nltk.NaiveBayesClassifier.train(train_set)
print(f"Accuracy on test set: {nltk.classify.accuracy(classifier, test_set):.2%}") #turned into percent
print(f"Accuracy on dev-test set: {nltk.classify.accuracy(classifier, devtest_set):.2%}") #turned into percent

Accuracy on test set: 78.60%
Accuracy on dev-test set: 75.60%


In [21]:
#Feature classifier 1 (Add first letter and last letter)
def gender_features1(word):
    return {'first_letter': word[0].lower(),
            'last_letter': word[-1].lower()}

#converted to feature sets
train_set1 = [(gender_features1(n), g) for (n, g) in train_names]
devtest_set1 = [(gender_features1(n), g) for (n, g) in devtest_names]
test_set1 = [(gender_features1(n), g) for (n, g) in test_names]

#obtained accuracy using naive Bayes classifier
classifier1=nltk.NaiveBayesClassifier.train(train_set1)
print(f"Accuracy on test set 1: {nltk.classify.accuracy(classifier1, test_set1):.2%}")
print(f"Accuracy on dev-test set 1: {nltk.classify.accuracy(classifier1, devtest_set1):.2%}")

Accuracy on test set 1: 77.40%
Accuracy on dev-test set 1: 78.20%


In [22]:
#Feature classifier 2 (Add first letter, one suffix, and last letter)
def gender_features2(word):
    return {'first_letter': word[0].lower(),
            'last_letter': word[-1].lower(),
            'suffix1':word[-1:].lower(),
            'suffix2': word[-2:].lower()}

#converted to feature sets
train_set2 = [(gender_features2(n), g) for (n, g) in train_names]
devtest_set2 = [(gender_features2(n), g) for (n, g) in devtest_names]
test_set2 = [(gender_features2(n), g) for (n, g) in test_names]

#obtained accuracy using naive Bayes classifier
classifier2=nltk.NaiveBayesClassifier.train(train_set2)
print(f"Accuracy on test set 2: {nltk.classify.accuracy(classifier2, test_set2):.2%}")
print(f"Accuracy on dev-test set 2: {nltk.classify.accuracy(classifier2, devtest_set2):.2%}")

Accuracy on test set 2: 78.00%
Accuracy on dev-test set 2: 76.40%


In [23]:
#Feature classifier 3 (Kitchen sink approach -add first letter, last letter, name length, and two suffix)
def gender_features3(word):
  word = word.lower()
  return {'first_letter': word[0],
            'last_letter': word[-1],
            'suffix1':word[-1:],
            'suffix2': word[-2:],
            'name_length' :len(word)}

#converted to feature sets
train_set3 = [(gender_features3(n), g) for (n, g) in train_names]
devtest_set3 = [(gender_features3(n), g) for (n, g) in devtest_names]
test_set3 = [(gender_features3(n), g) for (n, g) in test_names]

#obtained accuracy using naive Bayes classifier
classifier3=nltk.NaiveBayesClassifier.train(train_set3)
print(f"Accuracy on test set 3: {nltk.classify.accuracy(classifier3, test_set3):.2%}")
print(f"Accuracy on dev-test set 3: {nltk.classify.accuracy(classifier3, devtest_set3):.2%}")

Accuracy on test set 3: 77.80%
Accuracy on dev-test set 3: 77.00%


**Errors**

In [24]:
#generated a list of the errors that the classifier makes when predicting name genders
errors = []

for (name, tag) in devtest_names:
    guess = classifier.classify(gender_features(name))
    if guess != tag:
        errors.append((tag, guess, name))

# Print some errors
for (tag, guess, name) in errors[:20]:
    print(f"Correct: {tag:<6} Guess: {guess:<6} Name: {name}")

Correct: female Guess: male   Name: Leann
Correct: female Guess: male   Name: Jordain
Correct: male   Guess: female Name: Shorty
Correct: female Guess: male   Name: Vin
Correct: female Guess: male   Name: Kristyn
Correct: male   Guess: female Name: Pryce
Correct: female Guess: male   Name: Ingeborg
Correct: male   Guess: female Name: Lawerence
Correct: female Guess: male   Name: Mildrid
Correct: male   Guess: female Name: Yardley
Correct: female Guess: male   Name: Rakel
Correct: female Guess: male   Name: Storm
Correct: male   Guess: female Name: Burnaby
Correct: female Guess: male   Name: Devan
Correct: male   Guess: female Name: Dougie
Correct: male   Guess: female Name: Alfie
Correct: male   Guess: female Name: Jerrome
Correct: female Guess: male   Name: Erin
Correct: female Guess: male   Name: Caroljean
Correct: male   Guess: female Name: Dmitri


In [25]:
#generated a list of the errors that the classifier makes when predicting name genders
errors = []

for (name, tag) in devtest_names:
    guess = classifier1.classify(gender_features1(name))
    if guess != tag:
        errors.append((tag, guess, name))

# Print some errors
for (tag, guess, name) in errors[:20]:
    print(f"Correct: {tag:<6} Guess: {guess:<6} Name: {name}")

Correct: male   Guess: female Name: Kaleb
Correct: female Guess: male   Name: Jordain
Correct: male   Guess: female Name: Shorty
Correct: male   Guess: female Name: Abel
Correct: female Guess: male   Name: Hayley
Correct: male   Guess: female Name: Pryce
Correct: female Guess: male   Name: Ingeborg
Correct: male   Guess: female Name: Vern
Correct: male   Guess: female Name: Lawerence
Correct: female Guess: male   Name: Mildrid
Correct: male   Guess: female Name: Yardley
Correct: female Guess: male   Name: Rakel
Correct: female Guess: male   Name: Storm
Correct: male   Guess: female Name: Burnaby
Correct: female Guess: male   Name: Devan
Correct: female Guess: male   Name: Shirah
Correct: male   Guess: female Name: Dougie
Correct: male   Guess: female Name: Alfie
Correct: male   Guess: female Name: Jerrome
Correct: male   Guess: female Name: Austen


In [26]:
#generated a list of the errors that the classifier makes when predicting name genders
errors = []

for (name, tag) in devtest_names:
    guess = classifier2.classify(gender_features2(name))
    if guess != tag:
        errors.append((tag, guess, name))

# Print some errors
for (tag, guess, name) in errors[:20]:
    print(f"Correct: {tag:<6} Guess: {guess:<6} Name: {name}")

Correct: female Guess: male   Name: Jordain
Correct: male   Guess: female Name: Shorty
Correct: female Guess: male   Name: Josey
Correct: female Guess: male   Name: Vin
Correct: female Guess: male   Name: Hayley
Correct: female Guess: male   Name: Bamby
Correct: male   Guess: female Name: Pryce
Correct: female Guess: male   Name: Ingeborg
Correct: male   Guess: female Name: Lawerence
Correct: female Guess: male   Name: Mildrid
Correct: female Guess: male   Name: Rakel
Correct: female Guess: male   Name: Storm
Correct: female Guess: male   Name: Devan
Correct: female Guess: male   Name: Shirah
Correct: male   Guess: female Name: Dougie
Correct: male   Guess: female Name: Alfie
Correct: male   Guess: female Name: Jerrome
Correct: female Guess: male   Name: Erin
Correct: female Guess: male   Name: Caroljean
Correct: male   Guess: female Name: Dmitri


In [27]:
#generated a list of the errors that the classifier makes when predicting name genders
errors = []

for (name, tag) in devtest_names:
    guess = classifier3.classify(gender_features3(name))
    if guess != tag:
        errors.append((tag, guess, name))

# Print some errors
for (tag, guess, name) in errors[:20]:
    print(f"Correct: {tag:<6} Guess: {guess:<6} Name: {name}")

Correct: female Guess: male   Name: Jordain
Correct: male   Guess: female Name: Shorty
Correct: female Guess: male   Name: Josey
Correct: female Guess: male   Name: Vin
Correct: female Guess: male   Name: Hayley
Correct: female Guess: male   Name: Bamby
Correct: male   Guess: female Name: Pryce
Correct: female Guess: male   Name: Ingeborg
Correct: male   Guess: female Name: Lawerence
Correct: female Guess: male   Name: Mildrid
Correct: female Guess: male   Name: Rakel
Correct: female Guess: male   Name: Storm
Correct: female Guess: male   Name: Devan
Correct: female Guess: male   Name: Shirah
Correct: male   Guess: female Name: Dougie
Correct: male   Guess: female Name: Alfie
Correct: male   Guess: female Name: Jerrome
Correct: female Guess: male   Name: Erin
Correct: female Guess: male   Name: Caroljean
Correct: male   Guess: female Name: Dmitri


**Conclusion**
<br>In this assignment, I built a gender classifier for names using the NLTK Names Corpus and a Naive Bayes classifier. I split the data into a training set with 6,900 names, a dev-test set with 500 names, and a test set with 500 names. I started with a baseline feature (last letter of the name), I incrementally added features such as the first letter, suffixes, and name length.
The baseline classifier achieved approximately 78.6% accuracy on the test set and 75.6% on the dev-test set. Adding the first letter slightly improved consistency, with accuracies of 77.4% (test) and 78.2% (dev-test). Using the "kitchen sink" approach with multiple features, the classifier reached 77.8% (test) and 77.0% (dev-test). The kitchen sink approach may have worst due to overfitting.
<br>Overall, the classifier performed similarly on the dev-test and test sets, which is expected because both sets are random samples from the same distribution. Small differences in accuracy reflect natural variability in which specific names were included. The incremental improvements showed that simple features like first and last letters, suffixes, and name length capture much of the gender-related patterns in names, but some names remain ambiguous, which accounts for classification errors.